File: apps/android/app/src/main/kotlin/com/loonext/android/features/settings/VoiceGreetingCard.kt
package com.loonext.android.features.settings

import android.Manifest
import android.content.pm.PackageManager
import android.media.MediaPlayer
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import com.loonext.android.core.i18n.AppStrings
import com.loonext.android.core.i18n.LocalAppLocale
import com.loonext.android.core.i18n.t
import com.loonext.android.ui.common.userMessage
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * #309 — "Your own voice".
 *
 * Hand-port of `apps/web/src/components/settings/voice-greeting-card.tsx`,
 * keeping the three rules that shape it:
 *
 * - **You cannot save a take you have not heard.** Save appears only once
 *   there IS a recording, beside the player. Recording your own voice is the
 *   one thing people redo, and a flow that saves the first take unheard is one
 *   that assumes it was good.
 * - **The name is pre-filled** with what most owners are recording. An empty
 *   required field between somebody and their first playback is friction with
 *   no purpose.
 * - **Deleting asks first**, because it changes what every caller to a line
 *   using it hears and this card cannot show which lines those are.
 * - **There is a second way in, and it is a phone call.** Some owners will
 *   never record in an app — the permission prompt, the phone held at arm's
 *   length. "Have us call you" rings them and they talk. It sits behind a text
 *   button rather than beside Record, because two equally-weighted ways to do
 *   one thing is a decision nobody asked for, and it moves to the front the
 *   moment the microphone is refused, which is exactly when it is the answer.
 */
@Composable
internal fun VoiceGreetingCard(scope: SettingsScope, canEdit: Boolean) {
    val context = LocalContext.current
    val coroutines = rememberCoroutineScope()
    // The recorder's failures and the save confirmations are all written from
    // callbacks and coroutines, so the reader's language is read once here.
    val locale = LocalAppLocale.current
    val recorder = remember { GreetingRecorder(context) }

    var rows by remember { mutableStateOf(emptyList<VoicemailGreeting>()) }
    var take by remember { mutableStateOf<GreetingTake?>(null) }
    var name by remember { mutableStateOf(DEFAULT_GREETING_NAME) }
    var recording by remember { mutableStateOf(false) }
    var pending by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }
    var confirmDelete by remember { mutableStateOf<VoicemailGreeting?>(null) }
    // #309's phone path. `null` is closed; `phase` tells the one dialog whether
    // it is still asking or already waiting on a call that is out there.
    var capture by remember { mutableStateOf<CaptureState?>(null) }
    // True once the microphone has actually been refused, which is when the
    // phone path stops being an alternative and starts being the way through.
    var micRefused by remember { mutableStateOf(false) }
    // The take's own player. Held here so it can be released with the screen —
    // a MediaPlayer left holding a file is a leak the member never sees.
    val player = remember { mutableStateOf<MediaPlayer?>(null) }

    suspend fun refresh() {
        rows = try {
            scope.repo.voicemailGreetings(scope.companyId)
        } catch (cause: Exception) {
            error = cause.userMessage()
            rows
        }
    }

    LaunchedEffect(scope.companyId) { refresh() }

    // A microphone left open by a screen the member walked away from is the one
    // thing this must never do.
    DisposableEffect(Unit) {
        onDispose {
            recorder.discard()
            player.value?.release()
            player.value = null
        }
    }

    /**
     * Hear it back.
     *
     * This is not a nicety: the whole flow rests on an owner hearing their own
     * take before it becomes what customers hear. Without a player here the
     * Android card would ask them to trust a recording they have never heard,
     * which is the failure the web version was built to avoid.
     */
    fun playTake() {
        val current = take ?: return
        player.value?.release()
        player.value = try {
            MediaPlayer().apply {
                setDataSource(current.file.absolutePath)
                setOnCompletionListener { it.release(); player.value = null }
                prepare()
                start()
            }
        } catch (_: Exception) {
            error = AppStrings.translate(locale, "settingsMore.takeWontPlay")
            null
        }
    }

    fun begin() {
        error = null
        if (!recorder.start()) {
            // Almost always the mic is held by a call, or the permission was
            // revoked between the check and here. Say what to do, not that it
            // failed.
            error = AppStrings.translate(locale, "settingsMore.micUnavailable")
            micRefused = true
            return
        }
        micRefused = false
        recording = true
    }

    val micLauncher = rememberLauncherForActivityResult(
        ActivityResultContracts.RequestPermission(),
    ) { granted ->
        if (granted) {
            begin()
        } else {
            error = AppStrings.translate(locale, "settingsMore.micRefused")
            micRefused = true
        }
    }

    fun onRecord() {
        val granted = context.checkSelfPermission(Manifest.permission.RECORD_AUDIO) ==
            PackageManager.PERMISSION_GRANTED
        if (granted) begin() else micLauncher.launch(Manifest.permission.RECORD_AUDIO)
    }

    fun onStop() {
        recording = false
        take = recorder.finish()
        if (take == null) {
            error = AppStrings.translate(locale, "settingsMore.nothingRecorded")
        }
    }

    fun onSave() {
        val current = take ?: return
        coroutines.launch {
            pending = true
            error = null
            try {
                scope.repo.recordGreeting(
                    scope.companyId,
                    name.trim(),
                    current.durationMs,
                    current.bytes,
                )
                player.value?.release()
                player.value = null
                recorder.discard()
                take = null
                name = DEFAULT_GREETING_NAME
                refresh()
                scope.showMessage(
                    AppStrings.translate(locale, "settingsMore.greetingSaved"),
                )
            } catch (cause: Exception) {
                error = cause.userMessage()
            } finally {
                pending = false
            }
        }
    }

    /**
     * The greeting landing in the list IS the end of the phone flow.
     *
     * The owner is on a call and away from this screen, so the only
     * confirmation the call can produce is the row appearing. Polled by NAME
     * rather than by count: a second person recording at the same moment would
     * move a count and mean nothing about this call.
     */
    LaunchedEffect(capture?.phase, capture?.name) {
        val waiting = capture ?: return@LaunchedEffect
        if (waiting.phase != CapturePhase.CALLING) return@LaunchedEffect
        repeat(CAPTURE_POLLS) {
            delay(CAPTURE_POLL_MS)
            refresh()
            if (rows.any { it.name == waiting.name }) {
                capture = null
                scope.showMessage(
                    AppStrings.translate(
                        locale,
                        "settingsMore.namedGreetingSaved",
                        mapOf("name" to waiting.name),
                    ),
                )
                return@LaunchedEffect
            }
        }
    }

    fun startCaptureCall() {
        val current = capture ?: return
        coroutines.launch {
            pending = true
            error = null
            try {
                scope.repo.greetingCaptureCall(
                    scope.companyId,
                    current.name.trim(),
                    current.to.trim(),
                )
                capture = current.copy(phase = CapturePhase.CALLING)
            } catch (cause: Exception) {
                error = cause.userMessage()
            } finally {
                pending = false
            }
        }
    }

    SettingsCard(
        title = t("settingsMore.ownVoiceTitle"),
        description = t("settingsMore.ownVoiceDesc"),
    ) {
        Text(
            if (rows.isEmpty()) {
                t("settingsMore.noGreetingsYet")
            } else {
                t("settingsMore.pickGreetingOnNumber")
            },
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )

        rows.forEach { row ->
            Row(
                Modifier.fillMaxWidth().padding(top = 10.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(row.name, style = MaterialTheme.typography.bodyMedium)
                Spacer(Modifier.width(10.dp))
                Text(
                    formatGreetingDuration(row.duration_ms),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                Spacer(Modifier.weight(1f))
                if (canEdit) {
                    TextButton(enabled = !pending, onClick = { confirmDelete = row }) {
                        Text(t("common.delete"))
                    }
                }
            }
        }

        if (canEdit) {
            Column(Modifier.padding(top = 14.dp)) {
                when {
                    take != null -> {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Text(
                                t(
                                    "settingsMore.recordedLength",
                                    "length" to formatGreetingDuration(take!!.durationMs),
                                ),
                                style = MaterialTheme.typography.bodySmall,
                            )
                            Spacer(Modifier.weight(1f))
                            TextButton(enabled = !pending, onClick = { playTake() }) {
                                Text(t("settingsMore.hearItBack"))
                            }
                        }
                        Text(
                            t("settingsMore.exactlyWhatCallerGets"),
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                        OutlinedTextField(
                            value = name,
                            onValueChange = { name = it },
                            label = { Text(t("settingsMore.nameIt")) },
                            singleLine = true,
                            enabled = !pending,
                            modifier = Modifier.fillMaxWidth().padding(top = 8.dp),
                        )
                        Row(Modifier.padding(top = 8.dp)) {
                            TextButton(
                                enabled = !pending,
                                onClick = {
                                    player.value?.release()
                                    player.value = null
                                    recorder.discard()
                                    take = null
                                },
                            ) { Text(t("settingsMore.recordAgain")) }
                            Spacer(Modifier.weight(1f))
                            Button(
                                enabled = !pending && name.isNotBlank(),
                                onClick = { onSave() },
                            ) {
                                Text(
                                    if (pending) {
                                        t("common.saving")
                                    } else {
                                        t("settingsMore.saveGreeting")
                                    },
                                )
                            }
                        }
                    }

                    recording -> Row(verticalAlignment = Alignment.CenterVertically) {
                        Text(
                            t("settingsMore.recordingNow"),
                            style = MaterialTheme.typography.bodyMedium,
                        )
                        Spacer(Modifier.weight(1f))
                        Button(onClick = { onStop() }) { Text(t("settingsMore.stop")) }
                    }

                    else -> Row(verticalAlignment = Alignment.CenterVertically) {
                        Text(
                            t("settingsMore.upToTwoMinutes"),
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                        Spacer(Modifier.weight(1f))
                        Button(onClick = { onRecord() }) { Text(t("settingsMore.record")) }
                    }
                }

                if (take == null && !recording) {
                    TextButton(
                        enabled = !pending,
                        onClick = {
                            capture = CaptureState(
                                phase = CapturePhase.FORM,
                                name = DEFAULT_GREETING_NAME,
                                to = "",
                            )
                        },
                    ) {
                        Text(
                            if (micRefused) {
                                t("settingsMore.haveUsCallYou")
                            } else {
                                t("settingsMore.ratherOnThePhone")
                            },
                        )
                    }
                }
            }
        }

        error?.let {
            Text(
                it,
                modifier = Modifier.padding(top = 8.dp),
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.error,
            )
        }
    }

    confirmDelete?.let { target ->
        AlertDialog(
            onDismissRequest = { confirmDelete = null },
            title = {
                Text(t("settingsMore.deleteGreetingTitle", "name" to target.name))
            },
            text = { Text(t("settingsMore.deleteGreetingBody")) },
            confirmButton = {
                TextButton(
                    enabled = !pending,
                    onClick = {
                        confirmDelete = null
                        coroutines.launch {
                            pending = true
                            error = null
                            try {
                                scope.repo.deleteGreeting(scope.companyId, target.id)
                                refresh()
                                scope.showMessage(
                                    AppStrings.translate(
                                        locale,
                                        "settingsMore.deletedToast",
                                    ),
                                )
                            } catch (cause: Exception) {
                                error = cause.userMessage()
                            } finally {
                                pending = false
                            }
                        }
                    },
                ) { Text(t("common.delete")) }
            },
            dismissButton = {
                TextButton(onClick = { confirmDelete = null }) {
                    Text(t("settingsMore.keepIt"))
                }
            },
        )
    }

    capture?.let { state ->
        if (state.phase == CapturePhase.CALLING) {
            AlertDialog(
                onDismissRequest = { capture = null },
                title = { Text(t("settingsMore.callingNow", "number" to state.to)) },
                text = {
                    Column {
                        Text(t("settingsMore.answerAndListen"))
                        Spacer(Modifier.height(8.dp))
                        Text(t("settingsMore.captureStep1"))
                        Text(t("settingsMore.captureStep2"))
                        Text(t("settingsMore.captureStep3"))
                        Spacer(Modifier.height(8.dp))
                        Text(
                            t("settingsMore.captureWillAppear", "name" to state.name),
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                },
                confirmButton = {
                    TextButton(onClick = { capture = null }) { Text(t("common.close")) }
                },
            )
        } else {
            AlertDialog(
                onDismissRequest = { capture = null },
                title = { Text(t("settingsMore.recordOnPhone")) },
                text = {
                    Column {
                        Text(t("settingsMore.recordOnPhoneBody"))
                        OutlinedTextField(
                            value = state.to,
                            onValueChange = { capture = state.copy(to = it) },
                            label = { Text(t("settingsMore.yourNumber")) },
                            placeholder = { Text(t("settingsMore.captureNumberSample")) },
                            singleLine = true,
                            enabled = !pending,
                            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Phone),
                            modifier = Modifier.fillMaxWidth().padding(top = 12.dp),
                        )
                        OutlinedTextField(
                            value = state.name,
                            onValueChange = { capture = state.copy(name = it) },
                            label = { Text(t("settingsMore.nameIt")) },
                            singleLine = true,
                            enabled = !pending,
                            modifier = Modifier.fillMaxWidth().padding(top = 8.dp),
                        )
                    }
                },
                confirmButton = {
                    TextButton(
                        enabled = !pending && state.to.isNotBlank() && state.name.isNotBlank(),
                        onClick = { startCaptureCall() },
                    ) {
                        Text(
                            if (pending) {
                                t("settingsMore.calling")
                            } else {
                                t("settingsMore.callMe")
                            },
                        )
                    }
                },
                dismissButton = {
                    TextButton(onClick = { capture = null }) { Text(t("common.cancel")) }
                },
            )
        }
    }
}

/** Where the phone flow is: still asking, or already on a call. */
internal enum class CapturePhase { FORM, CALLING }

/**
 * The phone flow's whole state.
 *
 * One dialog holds both phases because they are one errand — an owner who puts
 * the phone to their ear mid-flow comes back to the window that told them what
 * to do, not to a closed one.
 */
internal data class CaptureState(
    val phase: CapturePhase,
    val name: String,
    val to: String,
)

/** Five seconds apart for three minutes: long enough to cover a 45-second ring,
 *  a two-minute recording, and the seconds it takes us to store it. */
private const val CAPTURE_POLL_MS = 5_000L
private const val CAPTURE_POLLS = 36

/**
 * What most owners are recording, so the field is never empty.
 *
 * Still editable — a workspace with a holiday greeting and a truck greeting
 * needs to say which is which — but nobody should have to think of a name
 * before they can hear their first take.
 */
private const val DEFAULT_GREETING_NAME = "After hours"

File: apps/ios/Loonext/Features/Settings/VoiceGreetingCard.swift
import SwiftUI

/**
 #309 — "Your own voice".

 Hand-port of `apps/web/src/components/settings/voice-greeting-card.tsx` and
 `VoiceGreetingCard.kt`, keeping the three rules that shape it:

 - **You cannot save a take you have not heard.** Save appears only once there
   IS a recording, beside the button that plays it. Recording your own voice is
   the one thing people redo, and a flow that saves the first take unheard is
   one that assumes it was good.
 - **The name is pre-filled** with what most owners are recording. An empty
   required field between somebody and their first playback is friction with no
   purpose.
 - **Deleting asks first**, because it changes what every caller to a line using
   it hears and this card cannot show which lines those are.
 - **There is a second way in, and it is a phone call.** Some owners will never
   record in an app — the permission prompt, the phone held at arm's length.
   "Have us call you" rings them and they talk. It sits behind a plain button
   rather than beside Record, because two equally-weighted ways to do one thing
   is a decision nobody asked for, and its words change the moment the
   microphone is refused, which is exactly when it is the answer.
 */
struct VoiceGreetingCard: View {
    let scope: SettingsScope
    let canEdit: Bool

    /// Built from the scope's own graph rather than passed in: this is the only
    /// multipart call in Settings, and threading a client through the section
    /// for one card would be a parameter every other card ignores.
    private var multipart: MultipartClient {
        MultipartClient(api: scope.graph.api, sessionStore: scope.graph.sessionStore)
    }

    @StateObject private var recorder = GreetingRecorder()
    @State private var rows: [VoicemailGreeting] = []
    @State private var take: GreetingTake?
    @State private var name = defaultGreetingName
    @State private var recording = false
    @State private var pending = false
    @State private var error: String?
    @State private var confirmDelete: VoicemailGreeting?
    /// #309's phone path. `nil` is closed; the phase tells the one sheet whether
    /// it is still asking or already waiting on a call that is out there.
    @State private var capture: CaptureState?
    /// True once the microphone has actually been refused, which is when the
    /// phone path stops being an alternative and starts being the way through.
    @State private var micRefused = false

    @Environment(\.appLocale) private var appLocale

    private func t(_ key: String, _ vars: [String: String] = [:]) -> String {
        AppStrings.translate(appLocale, key, vars)
    }

    var body: some View {
        SettingsCard(
            title: t("settingsMore.ownVoiceTitle"),
            description: t("settingsMore.ownVoiceDesc")
        ) {
            Text(
                rows.isEmpty
                    ? t("settingsMore.noGreetingsYet")
                    : t("settingsMore.pickGreetingOnNumber")
            )
            .font(.footnote)
            .foregroundStyle(.secondary)

            ForEach(rows) { row in
                HStack {
                    Text(row.name).font(.callout)
                    Text(formatGreetingDuration(row.duration_ms))
                        .font(.caption)
                        .foregroundStyle(.secondary)
                    Spacer()
                    if canEdit {
                        Button(t("common.delete")) { confirmDelete = row }
                            .font(.caption)
                            .disabled(pending)
                    }
                }
                .padding(.top, 10)
            }

            if canEdit { composer }

            InlineError(error)
        }
        .task { await refresh() }
        .onDisappear { recorder.discard() }
        .alert(
            t("settingsMore.deleteGreetingTitle", ["name": confirmDelete?.name ?? ""]),
            isPresented: Binding(
                get: { confirmDelete != nil },
                set: { if !$0 { confirmDelete = nil } }
            )
        ) {
            Button(t("settingsMore.keepIt"), role: .cancel) { confirmDelete = nil }
            Button(t("common.delete"), role: .destructive) {
                if let target = confirmDelete { remove(target) }
                confirmDelete = nil
            }
        } message: {
            Text(t("settingsMore.deleteGreetingBody"))
        }
        .sheet(
            isPresented: Binding(
                get: { capture != nil },
                set: { if !$0 { capture = nil } }
            )
        ) {
            captureSheet
        }
    }

    /// The phone path. One sheet, two states — asking, then waiting — because
    /// they are one errand: an owner who puts the phone to their ear mid-flow
    /// comes back to the screen that told them what to do, not to a closed one.
    @ViewBuilder
    private var captureSheet: some View {
        NavigationStack {
            Form {
                if capture?.phase == .calling {
                    Section {
                        Text(t("settingsMore.answerAndListen"))
                        Text(t("settingsMore.captureStep1"))
                        Text(t("settingsMore.captureStep2"))
                        Text(t("settingsMore.captureStep3"))
                    } header: {
                        Text(t("settingsMore.callingNow", ["number": capture?.to ?? ""]))
                    } footer: {
                        // The name rides in as `{name}` rather than being
                        // assembled around a pair of escaped quotes, which is
                        // what this used to do and the comment that stood here
                        // was right to be wary of: the quotation marks are now
                        // the catalogue's problem, and French gets « » without
                        // this file knowing about it.
                        Text(t("settingsMore.captureWillAppear", ["name": capture?.name ?? ""]))
                    }
                } else {
                    Section {
                        TextField(
                            t("settingsMore.yourNumber"),
                            text: Binding(
                                get: { capture?.to ?? "" },
                                set: { capture?.to = $0 }
                            )
                        )
                        .keyboardType(.phonePad)
                        .textContentType(.telephoneNumber)
                        TextField(
                            t("settingsMore.nameIt"),
                            text: Binding(
                                get: { capture?.name ?? "" },
                                set: { capture?.name = $0 }
                            )
                        )
                    } footer: {
                        Text(t("settingsMore.recordOnPhoneBody"))
                    }
                }
                InlineError(error)
            }
            .navigationTitle(
                capture?.phase == .calling
                    ? t("settingsMore.captureOnTheWay")
                    : t("settingsMore.recordOnPhone")
            )
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button(
                        capture?.phase == .calling ? t("common.close") : t("common.cancel")
                    ) { capture = nil }
                }
                if capture?.phase != .calling {
                    ToolbarItem(placement: .confirmationAction) {
                        Button(
                            pending ? t("settingsMore.calling") : t("settingsMore.callMe")
                        ) { startCaptureCall() }
                            .disabled(
                                pending
                                    || (capture?.to ?? "").trimmingCharacters(in: .whitespaces).isEmpty
                                    || (capture?.name ?? "").trimmingCharacters(in: .whitespaces).isEmpty
                            )
                    }
                }
            }
        }
    }

    @ViewBuilder
    private var composer: some View {
        VStack(alignment: .leading, spacing: 8) {
            if let current = take {
                HStack {
                    Text(
                        t(
                            "settingsMore.recordedLength",
                            ["length": formatGreetingDuration(current.durationMs)]
                        )
                    )
                    .font(.footnote)
                    Spacer()
                    Button(t("settingsMore.hearItBack")) {
                        if !recorder.play(current) {
                            error = t("settingsMore.takeWontPlay")
                        }
                    }
                    .font(.caption)
                    .disabled(pending)
                }
                Text(t("settingsMore.exactlyWhatCallerGets"))
                    .font(.caption)
                    .foregroundStyle(.secondary)
                TextField(t("settingsMore.nameIt"), text: $name)
                    .textFieldStyle(.roundedBorder)
                    .disabled(pending)
                HStack {
                    Button(t("settingsMore.recordAgain")) {
                        recorder.discard()
                        take = nil
                    }
                    .disabled(pending)
                    Spacer()
                    Button(
                        pending ? t("common.saving") : t("settingsMore.saveGreeting")
                    ) { save() }
                        .buttonStyle(.borderedProminent)
                        .tint(BrandColor.olive)
                        .disabled(pending || name.trimmingCharacters(in: .whitespaces).isEmpty)
                }
            } else if recording {
                HStack {
                    Text(t("settingsMore.recordingNow")).font(.callout)
                    Spacer()
                    Button(t("settingsMore.stop")) { stop() }
                        .buttonStyle(.borderedProminent)
                        .tint(BrandColor.olive)
                }
            } else {
                HStack {
                    Text(t("settingsMore.upToTwoMinutes"))
                        .font(.caption)
                        .foregroundStyle(.secondary)
                    Spacer()
                    Button(t("settingsMore.record")) { begin() }
                        .buttonStyle(.borderedProminent)
                        .tint(BrandColor.olive)
                }
            }
            if take == nil && !recording {
                Button(
                    micRefused
                        ? t("settingsMore.haveUsCallYou")
                        : t("settingsMore.ratherOnThePhone")
                ) {
                    error = nil
                    capture = CaptureState(phase: .form, name: defaultGreetingName, to: "")
                }
                .font(.caption)
                .disabled(pending)
            }
        }
        .padding(.top, 14)
    }

    private func refresh() async {
        do {
            rows = try await scope.repo.voicemailGreetings(scope.companyId)
        } catch {
            self.error = error.userMessage
        }
    }

    private func begin() {
        error = nil
        // `callInProgress` is false here rather than plumbed: this card lives in
        // Settings, and the recorder's own `.record` session activation is what
        // actually fails if a call holds the audio — which surfaces as
        // `.couldNotStart` with a message that says to try again.
        if let refusal = recorder.start(callInProgress: false) {
            error = refusal.message(appLocale)
            micRefused = true
            return
        }
        micRefused = false
        recording = true
    }

    private func stop() {
        recording = false
        take = recorder.finish()
        if take == nil {
            error = t("settingsMore.nothingRecorded")
        }
    }

    private func save() {
        guard let current = take else { return }
        Task { @MainActor in
            pending = true
            error = nil
            defer { pending = false }
            do {
                _ = try await multipart.postFile(
                    path: "/v1/voicemail-greetings",
                    companyId: scope.companyId,
                    fields: [
                        ("name", name.trimmingCharacters(in: .whitespaces)),
                        ("duration_ms", String(current.durationMs)),
                    ],
                    fileField: "file",
                    fileName: "greeting.m4a",
                    contentType: "audio/mp4",
                    bytes: current.audio
                )
                recorder.discard()
                take = nil
                name = defaultGreetingName
                await refresh()
            } catch {
                self.error = error.userMessage
            }
        }
    }

    private func startCaptureCall() {
        guard let current = capture, current.phase == .form else { return }
        let wanted = current.name.trimmingCharacters(in: .whitespaces)
        Task { @MainActor in
            pending = true
            error = nil
            do {
                try await scope.repo.greetingCaptureCall(
                    scope.companyId,
                    name: wanted,
                    to: current.to.trimmingCharacters(in: .whitespaces)
                )
                pending = false
                capture?.phase = .calling
                await awaitCapturedGreeting(named: wanted)
            } catch {
                pending = false
                self.error = error.userMessage
            }
        }
    }

    /// The greeting landing in the list IS the end of the phone flow.
    ///
    /// The owner is on a call and away from this screen, so the only
    /// confirmation the call can produce is the row appearing. Watched by NAME
    /// rather than by count: a second person recording at the same moment would
    /// move a count and mean nothing about this call.
    private func awaitCapturedGreeting(named wanted: String) async {
        for _ in 0..<capturePollCount {
            try? await Task.sleep(nanoseconds: capturePollNanos)
            if capture == nil { return }
            await refresh()
            if rows.contains(where: { $0.name == wanted }) {
                capture = nil
                return
            }
        }
    }

    private func remove(_ target: VoicemailGreeting) {
        Task { @MainActor in
            pending = true
            error = nil
            defer { pending = false }
            do {
                try await scope.repo.deleteGreeting(scope.companyId, id: target.id)
                await refresh()
            } catch {
                self.error = error.userMessage
            }
        }
    }
}

/**
 What most owners are recording, so the field is never empty.

 Still editable — a workspace with a holiday greeting and a truck greeting
 needs to say which is which — but nobody should have to think of a name before
 they can hear their first take.
 */
private let defaultGreetingName = "After hours"

/// Where the phone flow is: still asking, or already on a call.
enum CapturePhase {
    case form
    case calling
}

/// The phone flow's whole state, carried by the one sheet that shows both.
struct CaptureState {
    var phase: CapturePhase
    var name: String
    var to: String
}

/// Five seconds apart for three minutes: long enough to cover a 45-second ring,
/// a two-minute recording, and the seconds it takes us to store it.
private let capturePollNanos: UInt64 = 5_000_000_000
private let capturePollCount = 36
